components/stacks-network/src/orchestrator.rs

```diff
@@ -2692,105 +2901,166 @@ events_keys = ["*"]
-        loop {
-            let rpc_call = base_builder(
-                &bitcoin_node_url,
-                &devnet_config.bitcoin_node_username,
-                &devnet_config.bitcoin_node_password,
-            )
-            .json(&json!({
+        // Only generate blocks if we're NOT using cached data
+        if !using_cache {
+            let _ = devnet_event_tx.send(DevnetEvent::info(
+                "Initializing blockchain with fresh blocks".to_string(),
+            ));
+            let mut error_count = 0;
+            loop {
+                let rpc_call = base_builder(
+                    &bitcoin_node_url,
+                    &devnet_config.bitcoin_node_username,
+                    &devnet_config.bitcoin_node_password,
+                )
+                .json(&json!({
+                    "jsonrpc": "1.0",
+                    "id": "stacks-network",
+                    "method": "generatetoaddress",
+                    "params": [json!(3), json!(miner_address)]
+                }))
+                .send()
+                .await
+                .map_err(|e| format!("unable to send 'generatetoaddress' request ({})", e));
+
+                match rpc_call {
+                    Ok(_r) => break,
+                    Err(e) => {
+                        error_count += 1;
+                        if error_count > max_errors {
+                            return Err(e);
+                        } else if error_count > 1 {
+                            let _ = devnet_event_tx.send(DevnetEvent::error(e));
+                        }
+                    }
+                }
+                std::thread::sleep(std::time::Duration::from_secs(1));
+
+                let _ =
+                    devnet_event_tx.send(DevnetEvent::info("Waiting for bitcoin-node".to_string()));
+            }
+
+            let mut error_count = 0;
+            loop {
+                let rpc_call = base_builder(
+                    &bitcoin_node_url,
+                    &devnet_config.bitcoin_node_username,
+                    &devnet_config.bitcoin_node_password,
+                )
+                .json(&json!({
+                    "jsonrpc": "1.0",
+                    "id": "stacks-network",
+                    "method": "generatetoaddress",
+                    "params": [json!(97), json!(faucet_address)]
+                }))
+                .send()
+                .await
+                .map_err(|e| format!("unable to send 'generatetoaddress' request ({})", e));
+
+                match rpc_call {
+                    Ok(_r) => break,
+                    Err(e) => {
+                        error_count += 1;
+                        if error_count > max_errors {
+                            return Err(e);
+                        } else if error_count > 1 {
+                            let _ = devnet_event_tx.send(DevnetEvent::error(e));
+                        }
+                    }
+                }
```

Comment:
Optional change, but IMO `let`/`else` is a little cleaner here and results in less indentation (same thing at lines 2926, 2993, and 3029):

```suggestion
                let Err(e) = rpc_call else {
                    break;
                };
                error_count += 1;
                if error_count > max_errors {
                    return Err(e);
                } else if error_count > 1 {
                    let _ = devnet_event_tx.send(DevnetEvent::error(e));
                }
```